### PR TITLE
Fix retrieving autosaves when using a custom rest_namespace

### PR DESCRIPTION
**Pull request by tomjn**
## What?

When specifying `rest_namespace` in a custom post type, a hardcoded `wp/v2` in `getAutosaves` results in a 404

So instead, use the rest namespace from the type, not just the base

## Why?

Because bugs are bad and CPT's with `rest_namespace` deserve autosaves too!

## How?

By not hardcoding `/wp/v2`

## Testing Instructions

1. Register a CPT named `test`
2. In the registration, use `rest_namespace` and set it to something such as `testing/v99`
3. Save permalinks
4. Try to edit a post and there should not be a 404 for `wp/v2/test?autosaves....`  in the browser error console

### CODE REVIEW

**Comment by TimothyBJacobs** [comment]
Thanks for the PR @tomjn!

The test failures do look related. I'm guessing that the post types response is mocked somewhere and that mock is missing a `rest_namespace` field definition?

**Comment by adamsilverstein** [comment]
Good catch @tomjn - your code probably just needs a fallback if namespace is undefined to get tests passing. I did notice we have some other non-test instances of `wp/v2` being hard coded (for example post trashing also has it, see https://github.com/WordPress/gutenberg/search?q=wp%2Fv2). it would be good to:
1. fix all instances of hard-coded name spaces
2. add tests with post with custom namespaces to validate endpoints are constructed as expected

**Comment by tomjn** [comment]
I'll be honest, I wrote this in the github UI then took a flight so not surprised it failed a check or two, so this is my first chance to respond.

I do remember seeing the namespace referenced like this elsewhere:

```js
const namespace = postType?.rest_namespace ?? 'wp/v2';
```

**Comment by tomjn** [comment]
@adamsilverstein @TimothyBJacobs  I've addressed the linter issue, took me a bit more time than expected to get the dependencies/npm setup locally, I'm unsure on the best way to implement `??` while destructing. I thought about something like this:

```js
	const type = await resolveSelect.getPostType( postType );
	const base = type.rest_base;
	const namespace = type?.rest_namespace ?? 'wp/v2';
	const autosaves = await apiFetch( {
		path: `/${ namespace }/${ base }/${ postId }/autosaves?context=edit`,
	} );
```

I expect there'll be pushback over the naming of `type` though, suggestions?

**Comment by tomjn** [comment]
Updating the test mock data and adding new tests is beyond my bandwidth/knowledge for the project at the moment though, so I'll need help or someone else to pick it up

**Comment by TimothyBJacobs** [comment]
You should be able to assign a default while destructuring.

```js
  const { rest_namespace: restNamespace = 'wp/v2' } = resolveSelect.getPostType( postType );
```

**Comment by tomjn** [comment]
Screenshot 2022-06-07 at 11 08 54

🥳  thanks @TimothyBJacobs that has CI passing!

**Review by adamsilverstein — APPROVED** [review]
Looks good! Would be nice to add tests to validate and fix any other hard coded namespaces in a future PR.

**Comment by adamsilverstein** [comment]
Nice work @tomjn - approved! 

I'd love to see a follow up adding the tests and fixing any other hard coded namespace instances. Could you please create an issue to track this (even if you don't have time to work on the PR)?

**Comment by hgodinho** [comment]
Is there a way to disable the autosave feature until this PR is merged? 
The first call is taking more than 4 seconds and delaying the full loading of the page...

**Comment by tomjn** [comment]
@adamsilverstein  I no longer have the bandwidth to work on this, I don't have the capacity to push it forward, if you or someone else can create an issue for it I'd appreciate it, this was mostly me trying to help out someone on stack exchange and doesn't impact me personally. I also see there are now conflicts resulting from formatting changes made elsewhere

**Review by spacedmonkey — APPROVED** [review]
This looks good to go to me!

**Comment by spacedmonkey** [comment]
Related: https://github.com/WordPress/gutenberg/pull/41950

**Comment by Mamaduka** [comment]
Hi, @tomjn

It looks like static analysis checks are failing. Do you mind rebasing this branch and running the linter on the changed file?

**Comment by tomjn** [comment]
I don’t have the time to work on this, this will need a new owner to
champion it and push it forward

**Comment by Mamaduka** [comment]
Thanks for the quick reply, @tomjn. Happy to take over to PR and see this merged.

File: packages/core-data/src/resolvers.js [hunk 1]
```diff
@@ -348,11 +348,11 @@ export const canUserEditEntityRecord =
 export const getAutosaves =
 	( postType, postId ) =>
 	async ( { dispatch, resolveSelect } ) => {
-		const { rest_base: restBase } = await resolveSelect.getPostType(
+		const { rest_base: restBase, rest_namespace: restNamespace = 'wp/v2', } = await resolveSelect.getPostType(
 			postType
 		);
```
Comment (spacedmonkey) [inline packages/core-data/src/resolvers.js]:
```suggestion
		const { rest_base: restBase, rest_namespace: restNamespace = 'wp/v2' } =
			await resolveSelect.getPostType( postType );
```

**Comment by spacedmonkey** [comment]
@Mamaduka I think we are good to merge now.

**Comment by Mamaduka** [comment]
Thanks for the fix, @spacedmonkey!

**Pull request merged**